[REVAMP]: clean json mutation document (#479)
dgraph-io/dgraph-docs: content/dql-syntax/json-mutation-format.md

  parent = "dql-syntax"
  weight = 4
+++
Dgraph supports [Mutations]({{< relref "dql-mutation.md" >}}) in JSON or [RDF]({{< relref "dql-rdf.md" >}}) format.

> When using JSON format Dgraph creates nodes and relationships from the JSON structure and handles the assignation of UIDs to nodes.

Reviewer: ```suggestion
When using JSON format Dgraph creates nodes and relationships from the JSON structure and assigns UIDs to nodes.
```

## Specifying node UIDs

> For example, if you run the following mutation 

Reviewer: ```suggestion
For example, if you run this mutation: 
```

```dql
   {
   "set": [
     {
      "name": "diggy",
      "dgraph.type": "Mascot"
     }
   ]
 }  
```
You will see that Dgraph responds with 
{{< highlight json "linenos=false,hl_lines=7 " >}}
{
  "data": {
    "code": "Success",
    "message": "Done",
    "queries": null,
    "uids": {
      "dg.3162278161.22055": "0xfffd8d72745f0650"
    }
  }
{{< / highlight >}}

Meaning that Dgraph has created one node from the JSON. It has used the identifier `dg.3162278161.22055` during the transaction. And the final UID value for this node is `0xfffd8d72745f0650`.

> You can control the identifier name by specifying a `uid` field in your JSON data and using the notation : 

Reviewer: ```suggestion
You can control the identifier name by specifying a `uid` field in your JSON data and using the notation: 
```

``` "uid" : "_:<your-identifier>" ```

> In the following mutation we are using two JSON objects but as they are refering to the same identifier, Dgraph creates only one node :

Reviewer: ```suggestion
In this mutation, there are two JSON objects and because they are referring to the same identifier, Dgraph creates only one node :
```

```dql
   {
   "set": [
     {
      "uid": "_:diggy",
      "name": "diggy",
      "dgraph.type": "Mascot"
     },
     {
      "uid": "_:diggy",
      "specie": "badger"
     }
   ]
 }  
```

> Running this mutation, you can see that Dgraph returns the UID of the created node with the `diggy` identifier :

Reviewer: ```suggestion
When you run this mutation, you can see that Dgraph returns the UID of the node that was created with the `diggy` identifier:
```

{{< highlight json "linenos=false,hl_lines=7 " >}}
{
  "data": {
    "code": "Success",
    "message": "Done",
    "queries": null,
    "uids": {
      "diggy": "0xfffd8d72745f0691"
    }
{{< / highlight >}}

Note that the `specie` field is added to the node already created with `name` and `dgraph.type` information.

### Referencing existing nodes

You can use the `"uid"` field to reference an existing node. To do so, you must specify the UID value of the node.

For example:
```dql
   {
   "set": [
     {
      "uid": "0xfffd8d72745f0650",
      "specie": "badger"
     }
   ]
 }  
```

> Will add the specie information to the node we have created earlier.

Reviewer: ```suggestion
Adds the `specie` information to the node that was created earlier.
```

## Language support

To set a string value for a specific lnguage, append the language tag to the field name.

> Assuming that `specie` predicate has the @lang directive, the JSON mutation

Reviewer: ```suggestion
In case, `specie` predicate has the @lang directive, the JSON mutation
```

```dql
   {
   "set": [
     {
      "uid": "_:diggy",
      "name": "diggy",
      "dgraph.type": "Mascot",
      "specie@en" : "badger",
      "specie@fr" : "blaireau"
     }
   ]
 }  
```

> Will set the `specie` string predficate in two english and in french.

Reviewer: ```suggestion
Dgraph sets the `specie` string predicate in English and in French.
```

## Geolocation support

Geo-location data must be specified using keys `type` and `coordinates` in the JSON document. 
The supported types are `Point`, `Polygon`, or `MultiPolygon` . 

```dql
 {
   "set": [
     {
      "name": "diggy",
      "dgraph.type": "Mascot",
      "home" : {
          "type": "Point",
          "coordinates": [-122.475537, 37.769229 ]
       }
     }
   ]
 }    
```

## Relationships

Relationships are simply created from the nested structure of JSON.

For example:
```dql
 {
   "set": [
     {
      "uid": "_:diggy",
      "name": "diggy",
      "dgraph.type": "Mascot",
      "food" : [
        {
          "uid":"_:f1",
          "name": "earthworms"
        },
        {
          "uid":"_:f2",
          "name": "apples"
        }]
     }
   ]
 }

```

> Will result in the creation of three nodes and the `food` predicate will be a relationship.

Reviewer: ```suggestion
This result in the creation of three nodes and the `food` predicate as a relationship.
```

{{< highlight json "linenos=false,hl_lines=6-7 " >}}
{
  "data": {
    "code": "Success",
    "message": "Done",
    "queries": null,
    "uids": {
      "diggy": "0xfffd8d72745f06d7",
      "f1": "0xfffd8d72745f06d8",
      "f2": "0xfffd8d72745f06d9"
    }
    ...
{{< / highlight >}}

You can use references to existing nodes at any level of your nested JSON.

> ## Forbidden values

Reviewer: Do we need this heading? we don't seem to have an intro paragraph

Author: I moved the section at the end and named it "reserved values". We must mention it somewhere.

The string values `uid(...)`, `val(...)`are not accepted:  

## Deleting literal values

To delete node predicates, specify the UID of the node you are changing and set  
the predicates to delete to the JSON value `null`.

For example, to remove the predicate `name` from node  `0xfffd8d72745f0691` :
```dql
{
   "delete": [
     {
      "uid": "0xfffd8d72745f0691",
      "name": null
     }
   ]
}
```

## Deleting relationship

A relationhsip can be defined with a cardinality of 1 or many (list).
Setting a relationship to `null` removes all the relationships.

```JSON
{
  "uid": "0xfffd8d72745f06d7",
  "food": null
}
```

To delete a single relationhsip in a list, you must specify the target node of the relationship.

```dql
{
   "delete": [
      {
      "uid": "0xfffd8d72745f06d7",
      "food": {
          "uid": "0xfffd8d72745f06d9"
        }
      }
   ]
}

```

deletes only one `food` relationship.

> To delete all predicates of a given node :

Reviewer: ```suggestion
To delete all predicates of a given node:
```

- make sure the node has a `dgraph.type` predicate 
- the type is defined by the [Dgraph type system]({{< relref "type-system.md" >}})

> - run a delete mutation specifying only the uid field :

Reviewer: ```suggestion
- run a delete mutation specifying only the uid field
```

## Deleting Facets

> To delete a Facet, overwrite it :  when you run a mutation for the same entity without a facet, the existing facet will be deleted automatically.

Reviewer: ```suggestion
To delete a `Facet`, overwrite it. When you run a mutation for the same entity without a `Facet`, the existing `Facet` is deleted automatically.
```

## Handling arrays 

> To create a predicate as a list of string :

Reviewer: ```suggestion
To create a predicate as a list of string:
```

```JSON
{
   "set": [
    {
      "testList": [
        "Grape",
        "Apple",
        "Strawberry",
        "Banana",
        "watermelon"
      ]
    }
   ]
}
```

> Assuming `0x06` is the UID of the node created.

Reviewer: ```suggestion
For example, if  `0x06` is the UID of the node created.
```

> To remove one value from the list :

Reviewer: ```suggestion
To remove one value from the list:
```

}
```

> To add a value :

Reviewer: ```suggestion
To add a value:
```
